### DISCUSSION TITLE
Why `SELECT` after `UPDATE`?

### DISCUSSION THREAD

**Discussion by frederikhors** [ent/ent#2599] [Q&A] [answered]
Using this code:

```go
err := ent.WithTx(ctx, entClient, func(tx *generated.Tx) error {
  // other ops...
  
  entPlayer, err := tx.Player.UpdateOneID(id).SetName(newName).Save(ctx)
  // handle err

  // use entPlayer here...

  return nil
})
// handle err
```

I expect to have only the `UPDATE` SQL query.

Instead I get `UPDATE` and `SELECT`:

```
2022/06/04 16:14:34 driver.Tx(9d59627b-7d40-4cc6-b8cb-5e8b9c3748e4): started

2022/06/04 16:14:34 Tx(9d59627b-7d40-4cc6-b8cb-5e8b9c3748e4).Exec: query=UPDATE "players" SET "updated_at" = $1, "name" = $2 WHERE "id" = $3 AND "players"."tenant_id" = $4 args=[2022-06-04 16:14:34.3706568 John 1 2]

2022/06/04 16:14:34 Tx(9d59627b-7d40-4cc6-b8cb-5e8b9c3748e4).Query: query=SELECT "id", "created_at", "updated_at", "tenant_id", "name" FROM "players" WHERE "id" = $1 AND "players"."tenant_id" = $2 args=[1 2]

2022/06/04 16:14:34 Tx(9d59627b-7d40-4cc6-b8cb-5e8b9c3748e4): committed
```

**Why?**

I tried with `.Exec()` too but it's the same.

**Comment by a8m** [comment, marked as answer]
The `Exec` option uses `Save` under the hood and load the record - the same layer used by hooks/privacy of `UpdateOne` operation. If you don't want to load the record, use `Update().Where(player.ID(id))....Exec(ctx)` instead.

**Reply by frederikhors** [reply]
Thanks.

Using `.Update()` there is no SELECT but now `.Save()` does not return the saved object but only the number of nodes affected by the update operation.

Is there any way to have the whole object saved by UPDATE query without SELECT?

SQL uses "RETURNING *" for this usually.

**Reply by frederikhors** [reply]
Thinking about it better I can understand why `.Update()` doesn't return the object. However I would like to have a way to remove the SELECT on `.UpdateOne()/. UpdateOneID()` using `RETURNING`.

**Reply by frederikhors** [reply]
In a couple of tests against Bun (using the same driver) the updates are 150% slower.

**Reply by a8m** [reply]
Ent can use `RETURNING` in case the driver supports it (PostgreSQL and SQLite > 3.35.0). It requires some refactoring, and we already have other commitments for v0.11.

**Reply by frederikhors** [reply]
Yeah, but how? Is there a method? I can improve docs if you explain it to me.

**Reply by frederikhors** [reply]
Oh. Ok. Is not possible today. Ok. So in future the SELECT after the UPDATE will vanish, right?

**Reply by a8m** [reply]
Yes. The `sqlgraph` (some of Ent runtime code) was written before SQLite supported it, and this was a PostgreSQL thing. The change is not that big (a few if statements). I'll try to give it a look soon, but as mentioned, we already have commitments for v0.11 and we really want to close this release (95% done).

**Reply by frederikhors** [reply]
@a8m No problem. You are doing a Guinness job! :smile:

I ask you some questions just to UNDERSTAND. And mostly to give justification to the team (I'm the one pushing to use Ent).

I'll open an issue for this so we can track it, if you don't mind.

**Reply by rubensayshi** [reply]
@a8m is there any limitation why `Exec` needs to call `Save` under the hood?  
or should it be doable for me to make a PR that for `Exec` doesn't assign the `ScanValues` and as a result won't trigger a `SELECT` ?

**Reply by a8m** [reply]
Hey @rubensayshi! It's great to see you here again 😃 
The last time I experimented with it, I found that some mutation/policy code of ours (and other users) was relying on it. Here's one of them: https://github.com/ent/ent/pull/2112.

I still think the issue mentioned here should be addressed and added to the upcoming release.
